Add "trap" keyword for Go-like error handling

d-torrance opened the pull request:
This is an alternative to #4047 that may address @mahrud's concerns.  Rather than touching `try`, we add a new keyword `attempt` that returns a pair `(val, err)` (much like Go), where `val` is the value of the code and `err` is null if no error occurred, and otherwise `val` is null and `err` is an instance of a new `Error` type with information about the error that occurred .

```m2
i1 : attempt 5

o1 = (5, )

o1 : Sequence

i2 : attempt 1/0

o2 = (, Error{"message" => division by zero})
              "position" => stdio:2:8-2:11

o2 : Sequence
```

d-torrance committed:
Update JSONRPC to use "trap" for better error messages

Also bump to version 0.2 and start a changelog

d-torrance committed:
Allow Error objects to be sent to top level using SpecialExpr

Class is stored in "value" field so we can subclass

d-torrance committed:
Add "trap" keyword for go-like error handling

Returns a pair (value, err)

If code evaluates successfully, then err is null.  Otherwise, it's an
Error object with information about the error that occurred.

d-torrance committed:
Document and add tests for "trap" and "Error"

mahrud commented on M2/Macaulay2/packages/Cremona.m2:
```diff
@@ -1690,9 +1690,9 @@ point (Ideal,Boolean) := (I,b) -> (  -- see also: code(randomKRationalPoint,Idea
-       maxAttempts := 20; attempt := 0;
-       while #L == 0 and attempt < maxAttempts do (attempt = attempt+1; par = parametrize randomLinearSubspace(R,1); L = select(decompose par^* I,q -> dim q == 1 and degree q == 1));
-       if #L == 0 and attempt >= maxAttempts then error("reached maximum number of "|toString(maxAttempts)|" attempts to find point");
+       maxAttempts := 20; att := 0;
```
`tries` (and `maxTries`) would be better I think, let's not butcher their attempt at being readable (no pun intended lol)

mahrud commented:
I really like this solution!

seangrate commented:
This seems like a good approach. Hopefully at some point, we can also build in some sort of error types so that we can check for particular errors by their *types* as opposed to their *messages*.

d-torrance commented:
That sounds like a good idea!

pzinn commented:
what I don't understand is, the output of `attempt` is a pair, but either the first or the second element of the pair is `null`? what's the point? one might as well test if output is an instance of `Error`

pzinn commented:
the location of the error is different than in the usual error message. is that intentional?
```
i1 : 1/0
stdio:1:1:(3): error: division by zero

i2 : attempt(1/0)

o2 = (, Error{"message" => division by zero})
              "position" => stdio:2:8-2:11

o2 : Sequence
```
(it may not be obvious, but in the example the first error point to `/` whereas the second points to `1/0`)

pzinn commented:
This location doesn't seem right either:
```
i1 : attempt(1/a)

o1 = (, Error{"message" => no method for binary operator / applied to objects:})
                                       1 (of class ZZ)
                                 /     a (of class Symbol)
              "position" => ../../Macaulay2/m2/robust.m2:101:25-101:66

o1 : Sequence
```

d-torrance commented:
> what I don't understand is, the output of `attempt` is a pair, but either the first or the second element of the pair is `null`? what's the point? one might as well test if output is an instance of `Error`

That's a good point.  I was basically just copying the Go syntax that Mahrud pointed out in the other PR.  Also, in an earlier draft, the second element was a string, not an `Error` object, so it was necessary then.

d-torrance commented:
As for the file positions, as I recall, errors originally all start out with `dummyPosition` and then get updated in various places, right?  I think `attempt` is grabbing it a little later than it would have gotten printed if printing wasn't suppressed by `tryEval`, so maybe the position was updated a bit more in the meantime?

mahrud commented:
> what I don't understand is, the output of `attempt` is a pair, but either the first or the second element of the pair is `null`? what's the point? one might as well test if output is an instance of `Error`

Strictly speaking, in Go the return value is always of the expected type, so `attempt 1/0` should return `(0, Error{...})` or something like that, but currently there's just no output. But even for us, because of type consistency this is more readable:
```m2
(inv, err) := attempt inverse n
if err =!= null then error("oops: ", err#"message")
```
than a situation where the return value has different types in each branch:
```m2
ret := attempt inverse n
if not instance(ret, Error) then inv := ret else error("oops: ", ret#"message")
```

Also, what if you're writing an error-handling related function that needs to return an error? You need to separate the legitimate output Error object from the error that may have occurred in the process :)

pzinn commented:
I stand by my original point.

antonleykin commented on M2/Macaulay2/d/evaluate.d:
```diff
@@ -1670,6 +1670,22 @@ testfun(c:Code):Expr := (
+    else (
+	when r
+	is err:Error do seq(nullE, toExpr(err))
+	else seq(r, nullE))); -- shouldn't happen
```
This shouldn't happen, unless the code fabricates it's own `Error`. I don't think we should let `attempt` output user-fabricated stuff.

d-torrance replied:
Makes sense.  I've replaced this with an error.

antonleykin commented on M2/Macaulay2/packages/Macaulay2Doc/ov_debugging.m2:
```diff
@@ -327,6 +327,43 @@ document {
+    attempt c
+  Description
+    Text
+      The code @VAR "c"@ is evaluated and a sequence containing two elements
```
Emphasize that this is a keywork in the language, not a method. One could mentally parse `(attempt 1)/0`. 
Perhaps give more examples: what would `attempt 1/2; 1/0` give?

d-torrance replied:
Documentation expanded

antonleykin reviewed (changes requested):
The two ideas are good. 
The second (proposed by @seangrate) to have an `Error` type.  
The original idea to "capture message of error" (due to @d-torrance) via `attempt` (returning it, not storing, proposed by @mahrud).  

Returning one thing---either "result" or `Error`---would be more natural, since our language is not statically typed (e.g., the type of the "result" is unknown anyway).

d-torrance commented:
@antonleykin - I think I've addressed your comments.  `attempt` now returns a single element, either the value of the evaluated code or an `Error` object.

mahrud commented:
I would encourage you to reevaluate this decision. Here are a few dangers of this paradigm:
1. If a helper function `constructMyError` legitimately returns an `Error` object (we have LOTS of these helpers in Core), it's impossible to detect from `err := attempt constructMyError(...)` if `err` is a legitimate output or if `constructMyError` itself failed. We need Core to be robust, so I wouldn't use `attempt` there with this implementation.
2. Since error checking is less explicit, it's easy for users to forget to check for errors and only later end up with confusing type errors like `dim myFunc()` giving `error: no method found for applying dim to: argument : Error {...} (of class Error)`
3. Even worse, since `Error` inherits from hash tables, a caller can accidentally treat the output as input to a function that accepts a hash table and consequently mask the error. Here is an explicit example of this:
```m2
f := M -> attempt apply(codim M, i -> i^2);
# f(ZZ^100)      -- gives #{}         which is 0
# f((ZZ[x])^100) -- gives #Error{...} which is 2
```

If you're going through the trouble of adding a new keyword, I'd like it to be robust and future proof, especially for MPI or RPC type applications.

pzinn commented:
> I would encourage you to reevaluate this decision. Here are a few dangers of this paradigm:
> 
> 1. If a helper function `constructMyError` legitimately returns an `Error` object (we have LOTS of these helpers in Core), it's impossible to detect from `err := attempt constructMyError(...)` if `err` is a legitimate output or if `constructMyError` itself failed. We need Core to be robust, so I wouldn't use `attempt` there with this implementation.

this I don't understand at all. There seems to be a basic confusion here. These `Error` types are produced by the interpreter itself or at the level M2 level by calling `error`, *not* by creating an error manually. If we were to do what you suggest that would be a totally different usage which would require a separate discussion -- and a complete rewrite of this PR anyway.
> 3. Even worse, since `Error` inherits from hash tables, a caller can accidentally treat the output as input to a function that accepts a hash table and consequently mask the error. Here is an explicit example of this:
> 
> ```m2
> f := M -> attempt apply(codim M, i -> i^2);
> # f(ZZ^100)      -- gives #{}         which is 0
> # f((ZZ[x])^100) -- gives #Error{...} which is 2
> ```
well, if you're going to write `attempt`, that means you're expecting an error, so you should test for it, otherwise you probably shouldn't use `attempt` in the first place.
That being said, I agree that `Error` probably shouldn't inherit from `HashTable`; instead, just like `Dictionary`, it should be its own type.

mahrud commented:
> There seems to be a basic confusion here. These `Error` types are produced by the interpreter itself or at the level M2 level by calling `error`, _not_ by creating an error manually. If we were to do what you suggest that would be a totally different usage which would require a separate discussion

Why shouldn't error objects be produced at the top-level? Is something preventing it in this proposal? There is a huge upside, as @seangrate suggested above, in allowing custom subtypes inheriting from Error so that error checking looks more like `if instance(err, RingError) ...` rather than `if err#"message" == "incorrect ring" ...`. Otherwise why bother with a new type at all?

We currently don't have an Error type, but once it exists I guarantee this PR isn't going to be the end of it. Soon we'll have extensions like Python error types and Flint error types and use it with hooks and threads and loads more that we can't imagine now. (Side note: @d-torrance if `err` is an instance of `Error` then probably `error err` should produce an error with that location and message)

> well, if you're going to write attempt, that means you're expecting an error, so you should test for it, otherwise you probably shouldn't use attempt in the first place.

Mistakes like this will sneak in, either by people treating `attempt` like `try` or just mistakes in the programming where there are multiple branches and you forget to check if the output is an error in one branch.

More generally, I'm just surprised that we're going with a syntax like this:
```m2
c := attempt codim M;
if instance(c, Error) then
if c#"message" == "codim: expected an affine ring (consider Generic=>true to work over QQ"
then c = planZZ(M)
else error "unexpected module"
```

d-torrance commented:
I'm convinced by @mahrud's arguments and prefer the Go-style two-element output.

pzinn commented:
> 
> Why shouldn't error objects be produced at the top-level? Is something preventing it in this proposal? There is a huge upside, as @seangrate suggested above, in allowing custom subtypes inheriting from Error so that error checking looks more like `if instance(err, RingError) ...` rather than `if err#"message" == "incorrect ring" ...`. Otherwise why bother with a new type at all?

because that simply wouldn't work in the current setting. what would happen if you called your function that outputs an error your way without `attempt`? in fact, what's the role of `attempt` then? As I said, if you want to go this route, then we should scrap this PR and start from scratch with a completely different design.

There absolutely should be subtypes, but this is irrelevant to the point I'm making.
I'm simply suggesting `Error` be its own "root" type.

d-torrance committed:
Add error(Error) for raising errors from Error objects

d-torrance commented:
I went back to the Go-style output and also added some basic support for subclassing of `Error`:

```m2
i1 : MyError = new SelfInitializingType of Error;

i2 : MyError "foo"

o2 = MyError{"message" => foo}

o2 : MyError

i3 : error o2
stdio:3:5:(3): error: foo

i4 : attempt error o2

o4 = (, MyError{"message" => foo                            })
                "position" => ../m2/debugging.m2:21:22-21:75

o4 : Sequence
```

mahrud commented:
Looks like the error
> "position" => ../m2/debugging.m2:21:22-21:75

is pointing to this line:
```m2
error Error := err -> olderror new class err from processArgs err#"message"
```
No obligation to fix this in this PR, but obviously the location should point somewhere more appropriate. There are probably multiple ways to fix it (e.g. maybe moving `error(Error)` to the interpreter?), but one of my hopes from the `attempt` keyword is to fix these spurious error layers: https://github.com/Macaulay2/M2/issues/2568.

pzinn commented:
I understand less and less the logic. Now one can create `Error` objects at the M2 level, which are then fed to `error`, which at the d level proceeds to discard that `Error` object, only keeping track of the message, and creates a new `Error` object out of it... ??

mahrud commented:
The point of object oriented error handling is that instead of interrupting the computation, an error object is created to help the subsequent code resolve it and proceed without interruption. But there may be errors objects which can't be handled, so there should be a way to force an interrupt so that the user can intervene. This is where `error(Error)` comes in, to simplify this:
```m2
(c, err) := attempt codim M;
if instance(err, Error) then
if instance(err, RingError)
then c = planZZ M
else error("unexpected error: " | err#"message" | " at " | err#"position")
```
into this:
```m2
(c, err) := attempt codim M;
if instance(err, Error) then
if instance(err, RingError)
then c = planZZ M
else error(err) -- the original message and position are returned
```

d-torrance committed:
Add method for constructing Error objects from strings

d-torrance committed:
Improve toString(Error) now that we can see them at top level

d-torrance committed:
Add locate(Error) (also fix comment about location of d code)

d-torrance commented:
I made some changes that may address some of @pzinn's concerns.

`Error` is now a basic type rather than a type of hash table.  Under the hood, a top-level `Error` object is a `SpecialExpr` that contains the lower-level `Error` object.  So `error(Error)` just extracts that object rather than creating a new one.

Rather than getting the message and location by looking up keys in a hash table, we can use `toString` and `locate`, resp.

antonleykin commented:
I think we are getting caught in the weeds and should tread carefully as the discussion diverged
1. from a simple (but elegant) solution to having a construct that 
  - handles an exception/interrupt like `try` (not terminating the program)
  - but instead of `else` returning an `Error` object
2. to an extensive rethinking of how errors are done in Macaulay2

I would like the *language* modifications to be minimal unless we are constructing a new language (Macaulay3 or GoMacaulay). I believe we should encourage users to, as someone wise said,
"prefer mathematical methods to check the input first and from there run completely error free" or write a simple statement   `error "something is wrong"` when something unexpected happen. (In particular, we'd want people to not use `try`, `catch`, etc.)

Let's try to think about minimal changes of language could be done to accomplish 1. 
(The following is not a proposition to act upon. It has downsides.) 
Could `catch` be modified so that 
```
i2 : catch apply({-1,0,3}, i->1/i)                                                                             
stdio:2:26:(3):[1]: error: division by zero 
```
returns either a string that is printed or an `Error` object?

d-torrance commented:
I think it would be straightforward to use `catch` instead of `attempt` for this.  It would be a breaking change, though, as right now if you use `catch` on some code that raises an error, that error is raised.  So I'd prefer to stick with `attempt` (or some other new keyword).

I don't think we should return just a string, as otherwise `attempt "foo"` and `attempt error "foo"` would be indistinguishable.  And returning just the `Error` object is a problem too, as `attempt error "foo"` and `attempt new Error from "foo"` would also be indistinguishable.

@antonleykin - Is the "extensive rethinking of how errors are done in Macaulay2" the idea of subclassing `Error`?  That's something we get for essentially free as soon as we expose that class at top level.

antonleykin commented:
> I think it would be straightforward to use `catch` instead of `attempt` for this. It would be a breaking change, though, as right now if you use `catch` on some code that raises an error, that error is raised. 

@mikestillman  and I discussed this earlier today. In .m2 files, the keywords `catch` and `throw` appear to be used... 0 times. Perhaps we have a chance to repurpose (dead wood)?
   
> I don't think we should return just a string, as otherwise `attempt "foo"` and `attempt error "foo"` would be indistinguishable. And returning just the `Error` object is a problem too, as `attempt error "foo"` and `attempt new Error from "foo"` would also be indistinguishable.

This is a deficiency, but also seem to be a hypothetical edge case.
Suppose we are redesigning `catch`:  one can make 
* `throw` return an error object of one type (inheriting from say, `AbstractError`) 
* any error resulting from `error` return an error object of a distinct type
* force the code that attempts to return an error object to fail (interrupt the program) 

This would make returning a pair unnecessary. Returning a pair is, of course, not a big deal, but I think we'd be just paying tribute to the edge case (e.g., hypothetical users who would use nested `catch`es or micromanaging error objects at the top level). 

> @antonleykin - Is the "extensive rethinking of how errors are done in Macaulay2" the idea of subclassing `Error`? That's something we get for essentially free as soon as we expose that class at top level.

My point: we can export `Error` to top level and subclass it anyway we want, but I would stay away from advertising this to users, and also using it ourselves (at top level) unless absolutely necessary.

mahrud commented:
I have an alternative name suggestion: `trap` (because errors can't _escape_ lol)

> I would like the _language_ modifications to be minimal unless we are constructing a new language (Macaulay3 or GoMacaulay). I believe we should encourage users to, as someone wise said,
> "prefer mathematical methods to check the input first and from there run completely error free" or write a simple statement `error "something is wrong"` when something unexpected happen. (In particular, we'd want people to not use `try`, `catch`, etc.)

The emphasis there was on **mathematical methods**. Literally in the next sentence I said that RPC and MPI are robust protocols and need a robust implementation that covers every hypothetical edge case (and honestly I don't think the examples that Doug and I provided are hypothetical at all).

I wouldn't recommend a casual user to use `attempt` either, just like I wouldn't recommend they use `capture`, but both are robust methods which have an important place in the Core (like robust.m2 and debugging.m2) and specific other situations (like RPC/MPI/Python but also external software interfaces).

Also, I don't think anyone is suggesting an extensive rewrite of how errors are done in M2 today. I think this PR is ready to be merged right now. However, if we're introducing a new keyword, I'd like it to allow future extensions so we avoid a future breaking change.

ps: if you want to redefine `catch`, my vote is something more intuitive:
```m2
try (...)
catch   RingError do (...) -- do this if the ring is bad
catch ModuleError do (...) -- do this if the module is bad
else (...) -- do this if there's another error
```

pzinn commented on M2/Macaulay2/d/classes.dd:
```diff
@@ -116,6 +117,13 @@ export Parent(e:Expr):HashTable := (
+	if ancestor(x, errorClass) then x
+	else errorClass)
+    else errorClass);
+
```
can you explain this bit?

d-torrance replied:
This is kind of a hack, but this let us subclass `Error`.  We store the class in two places:

* In the `SpecialExpr` object
* In the `value` member of the low-level `Error` object, which is currently only used for things like `break`, `continue`, and `throw` that we wouldn't catch with `attempt` (since if they aren't handled yet, we'd get an "unhandled" error instead).

The reason we need it in both places is so that we don't lose the class information when we do `error(Error)`, which extracts the low-level `Error` object from the `SpecialExpr`.  So for example:

```m2
i1 : MyError = new SelfInitializingType of Error;

i2 : last attempt error MyError "foo"

o2 = foo

o2 : MyError
```

If we didn't store `MyError` inside the low-level `Error` object so that `attempt` can find it when it calls `toExpr(Error)`, then we'd lose the subclass information.

pzinn commented:
just trying to understand the latest changes. now this works:
```
new Error from 42
```
but what happens under the hood?
also why the need for a special `new Error from String`?

mahrud commented on M2/Macaulay2/tests/normal/error-messages.m2:
Might be worth adding a test for nested attempts, e.g. `attempt attempts error "foo"` gives `((, foo), )`

mahrud commented on M2/Macaulay2/m2/debugging.m2:
```diff
@@ -13,11 +13,14 @@ processArgs := args -> concatenate (
+error = method(Dispatch => Thing)
+error String :=
+error Error  := olderror
+error Thing  := olderror @@ processArgs
```
One of these is different lol
```m2
i1 : errorDepth=1;

i2 : error new Error from "hi"
stdio:2:5:(3): error: hi

i3 : error "hi"
stdio:3:5:(3): error: hi

i4 : error 12
Projects/M2/engine/M2/Macaulay2/m2/classes.m2:114:51:(1):[2]: error: 12
Projects/M2/engine/M2/Macaulay2/m2/classes.m2:114:51:(1): entering debugger (enter 'help' to see commands)
Projects/M2/engine/M2/Macaulay2/m2/classes.m2:114:50-114:55: --source code:
Function @@ Function := Function => (f,g) -> x -> f g x
```

Note that this shows up in unexpected places too:
```m2
i10 : ZZ^{1}
Projects/M2/engine/M2/Macaulay2/m2/classes.m2:114:51:(1):[3]: error: expected each multidegree to be of length 0
Projects/M2/engine/M2/Macaulay2/m2/classes.m2:114:51:(1): entering debugger (enter 'help' to see commands)
Projects/M2/engine/M2/Macaulay2/m2/classes.m2:114:50-114:55: --source code:
Function @@ Function := Function => (f,g) -> x -> f g x
```
At the very least, change `error(Thing)` to just `x -> olderror processArgs x`, then the interpreter stops at that line. But if you can make it so the interpreter never stops here that'd be super wonderful.

d-torrance replied:
Done!

```m2
i1 : errorDepth = 1;

i2 : error 12
../m2/debugging.m2:19:29:(1):[2]: error: 12
../m2/debugging.m2:19:29:(1): entering debugger (enter 'help' to see commands)
../m2/debugging.m2:19:21-19:43: --source code:
error Thing  := x -> olderror processArgs x

ii3 : ZZ^{1}
../m2/debugging.m2:19:29:(1):[3]: error: expected each multidegree to be of length 0
../m2/modules.m2:288:18:(1):[1]: --back trace--
```

d-torrance replied:
Hmm, the `ZZ^{1}` error isn't quite as useful if we break out of the debugger:

```m2
ii4 : break

i5 : ZZ^{1}
../m2/debugging.m2:19:29:(1):[3]: error: expected each multidegree to be of length 0
../m2/debugging.m2:19:29:(1): entering debugger (enter 'help' to see commands)
../m2/debugging.m2:19:21-19:43: --source code:
error Thing  := x -> olderror processArgs x
```

mahrud replied:
This is an epsilon better than the current behavior, which doesn't even fit the full function:
```m2
..../debugging.m2:18:5-18:30: --source code:
     olderror processArgs args)
```
But this annoyance is exactly why I suggested moving `error(Thing)` to the interpreter (or perhaps `startup.m2`? I will try this out at some point). I would love a general solution for #2568 that tells M2 "never interrupt at this point, it's irrelevant!" e.g. I don't want to step through methods and hooks and options while trying to figure out why my math function has an error. It's possible that `shield` was supposed to do this, but that hasn't worked for quite a while: #2596

d-torrance replied:
I went ahead and moved `error` completely to the interpreter:

```m2
i1 : errorDepth = 1

o1 = 1

i2 : ZZ^{1}
../m2/modules.m2:288:18:(1):[1]: error: expected each multidegree to be of length 0 (of class ZZ)
../m2/modules.m2:288:18:(1): entering debugger (enter 'help' to see commands)
../m2/modules.m2:288:13-288:71: --source code:
        then error("expected each multidegree to be of length ", degrk))
```

d-torrance replied:
Hmm, the `(of class ZZ)` is silly.  I used the same robust printing method that we were using for hash table key errors, where the class makes sense.  Here it doesn't.

mahrud replied:
Ha that can be confusing .. This is great though, thank you!

mahrud commented on M2/Macaulay2/d/evaluate.d:
```diff
@@ -1670,6 +1670,18 @@ testfun(c:Code):Expr := (
+	is err:Error do seq(nullE, toExpr(err))
+	-- shouldn't happen since tryEvalSuccess only false when r is an Error
+	else buildErrorPacket("unable to catch error")));
+setupop(attemptS, attemptfun);
```
Now that I look at how simple this is, could you remind me what's the issue or difficulty with this syntax?
```m2
try (...) catch err do (...) -- assigns the error object to `err`.
```
Importantly, based on the fact that `toExpr(err)` happens pretty late, I presume if we could use `catch` like this, normal `try .. else ..` would not be slowed down like the previous proposal. This might satisfy all parties (though part of me thinks a `trap` with a second error output would still be useful.

(If this doesn't work, I think renaming to `trap` will be good, so the error can say `unable to trap the error`. Also, the indentation looks strange. You could do `else when r is ...`)

d-torrance replied:
That's probably doable.  We'd probably need to introduce a new `TryCatchDo` member of the `ParseTree` union, but that shouldn't be too bad.  Would that be the only combination, or are you envisioning `else` and/or `then` showing up with `catch` and `do` as well? 

I am a little worried that `catch` may have too high of a precedence -- we'd probably want whatever keyword went here to have a low one matching `do`, `else`, and `then`.  Maybe `try ... trap ... do`?

mahrud replied:
I mean, as Anton pointed out catch isn't really used anywhere, so I think `try .. catch err do ..` makes more sense, but if you don't want to mess with precedences maybe `try .. except err do ..` instead? `trap` doesn't really make sense here.

mahrud replied:
I think we should limit the scope for now, but something like this could make sense in the future:
```m2
try ... then ...
catch   MyError as err do ...
catch RingError as err do ...
else ... -- for other errors
```

d-torrance replied:
I've switched it to `try ... except err do ...`

mahrud commented on M2/Macaulay2/d/actors4.d:
```diff
@@ -663,10 +663,13 @@ stringcatfun(e:Expr):Expr := (
-     else buildErrorPacket("expects a string or sequence of strings as its argument"));
+     is err:Error do (
+	 err.printed = false;
+	 e)
```
Sorry to be nitpicky but the indentation looks strange.
```suggestion
     is err:Error do ( err.printed = false; e )
```

d-torrance replied:
Fair enough.  I generally like to stick with one statement per line but `e` by itself is a little odd.  :) Fixed

d-torrance commented:
> but what happens under the hood? also why the need for a special `new Error from String`?

`new Error from String` is the low-level d constructor that constructs a new low-level `Error` object from the given message and stuffs it in a `SpecialExpr` to make the top-level `Error` object.

`new Error from Thing` is a top-level wrapper around this that calls the helper function `processArgs` on its input (which is also what we use when we call `error`) to construct a string that gets sent to the low-level method.

d-torrance committed:
Add "try ... except... do" statement for catching errors

d-torrance commented:
I went ahead and added a commit implementing a `try ... trap ... do` statement.  If this works well, then we can remove `attempt`.

```m2
i1 : try 1/0 trap err do err

o1 = division by zero

o1 : Error
```

mahrud commented:
Here's a suggestion to get things moving: let's add both `try ... except err do ...` and also `(out, err) = trap ...` (so you can undo renaming `attempt` in a few packages) in this PR, try both for a few months, and decide before the next release whether we want to keep one or both or make any improvements.

mahrud commented on M2/Macaulay2/d/parse.d:
```diff
@@ -197,6 +198,7 @@ export globalAssignmentCode := {+
      };
 export ifCode := {+ predicate:Code, thenClause:Code, elseClause:Code, position:Position };
 export tryCode := {+ code:Code, thenClause:Code, elseClause:Code, position:Position };
+export tryExceptDoCode := {+ code:Code, doClause:Code, frameID:int, framesize:int, position:Position };
```
Is there a reason this can't fall under the same `tryCode`? (for now we could flag an error if both `except` and `else` are given)

d-torrance replied:
I thought about that, but I think this keeps the code in `evaluate.d` a bit cleaner, where we don't have to worry about all the various cases of which clauses we have and don't have.

mahrud replied:
What about something like this? I feel like it's a pretty clean addition:
```m2
	  is c:tryCode do (
	      ret := tryEval(c.code);
	      if tryEvalSuccess then
	      -- what to do in case of success
	      when ret is Error do ret -- TODO: what kind of error leaves tryEvalSuccess=true?
	      else if c.thenClause == NullCode then ret   else eval(c.thenClause)
	      -- what to do in case of failure
	      else if c.doClause   != NullCode then evalExceptDoCode(c.doClause)
	      else if c.elseClause == NullCode then nullE else eval(c.elseClause))
```
where `evalExceptDoCode` is similar to `evalTryCode`.

mahrud replied:
Actually it should be `evalExceptDoCode(c.doClause, ret)`.

Do you know why only 6 types of errors are escape try but some don't? e.g.
```m2
i13 : step
stdio:13:0:(3): error: unhandled step command

i14 : break
stdio:14:0:(3): error: unhandled break command

i15 : try (1+1; step) else 3

o15 = 3

i16 : try (1+1; break) else 3
stdio:16:10:(3): error: unhandled break command
```
I gather for some types of errors we want an actual interruption (e.g. literal ctrl-c interrupt should not be captured by the else clause in a try) but idk why `step` is okay but `break` is bad.

d-torrance replied:
Good question -- I'm not sure.  `step` is only for the debugger, right?  Whereas `break` is used in both the debugger and the regular language, so maybe that's why?

Your `tryCode` suggestion seems reasonable -- I'll play around with it.

mahrud commented on M2/Macaulay2/d/convertr.d:
```diff
@@ -333,6 +333,7 @@ export convert0(e:ParseTree):Code := (
     is i:TryThen     do Code(tryCode(convert(i.primary), convert(i.sequel), NullCode,             pos))
     is i:TryThenElse do Code(tryCode(convert(i.primary), convert(i.sequel), convert(i.alternate), pos))
     is i:TryElse     do Code(tryCode(convert(i.primary), NullCode,          convert(i.alternate), pos))
+    is i:TryExceptDo do Code(tryExceptDoCode(convert(i.primary), convert(i.doClause), i.dictionary.frameID, i.dictionary.framesize, pos))
```
To elaborate on my previous comment, this could be something like:
```suggestion
    is i:TryExceptDo do Code(tryCode(convert(i.primary), NullCode, NullCode, convert(i.doClause), i.dictionary.frameID, i.dictionary.framesize, pos))
```
(And the other lines need to be adjusted for the new inputs too)

d-torrance committed:
Refactor RobustPrintMethod

Now takes a single input (the object to print) and leaves the
formatting of the message to the caller.

Also renamed to RobustPrintNetMethod so we can also have
RobustPrintSringMethod (which we'll use for arbitrary errors)

d-torrance committed:
Move toString(FilePosition) to interpreter

d-torrance committed:
Move "error" completely to the interpreter

d-torrance committed:
Document and add test for try/except/do

d-torrance commented:
I've made a few updates:

 * `attempt` is now `trap`
 * `tryExceptDoCode` has been removed and its functionality merged into `tryCode`
 * `RobustPrintMethod` has been split into two methods: `RobustPrintStringMethod` for printing arbitrary things passed to `error` and `RobustPrintNetMethod` for "key not found" errors.

mahrud commented on M2/Macaulay2/packages/Macaulay2Doc/ov_language.m2:
```diff
@@ -704,37 +704,70 @@ document {
+doc ///
+  Key
+    symbol try
+    "except" -- TODO: 'symbol except' here raises "global symbols inadvertently defined" error.  why?
```
Because it needs to be exported in `exports.m2`.

d-torrance replied:
D'oh -- thanks!

mahrud commented on M2/Macaulay2/d/convertr.d:
```diff
@@ -329,10 +329,11 @@ export convert0(e:ParseTree):Code := (
-    is i:TryThen     do Code(tryCode(convert(i.primary), convert(i.sequel), NullCode,             pos))
-    is i:TryThenElse do Code(tryCode(convert(i.primary), convert(i.sequel), convert(i.alternate), pos))
-    is i:TryElse     do Code(tryCode(convert(i.primary), NullCode,          convert(i.alternate), pos))
+    is i:Try         do Code(tryCode(convert(i.primary), NullCode,          NullCode,             NullCode, 0, 0, pos))
```
I think this is fine, but putting `frameID` 0 here isn't going to cause null pointer errors or weird access to other frames, right?

d-torrance replied:
We only ever use this value in the `try ... except ... do ... case` (or when just printing it in `pseudocode`/`disassemble`), so it should be fine

mahrud commented on M2/Macaulay2/d/evaluate.d:
```diff
@@ -1478,10 +1480,26 @@ export evalraw(c:Code):Expr := (
+	      else (
+		  if c.elseClause != NullCode then eval(c.elseClause)
+		  else if c.doClause == NullCode then nullE
+		  else ( -- try .. except .. do
```
Just pointing out that this is close to the suggestion I had, but different in one way: in my example, "except" overrides "else" but I think here "else" overrides "except". It may be a distinction without a difference in behavior, but worth be aware of.

d-torrance replied:
The parser will raise a syntax error if both are given, so I don't think it matters:

```m2
i1 : try 1/0 except err do err then "foo" else "bar"
stdio:1:26:(3): error: syntax error at 'then'

i1 : try 1/0 then "foo" else "bar" except err do err
stdio:2:30:(3): error: syntax error at 'except'
```

mahrud commented on M2/Macaulay2/d/evaluate.d:
```diff
@@ -1670,6 +1688,16 @@ testfun(c:Code):Expr := (
 
+trapfun(c:Code):Expr := (
+    r := tryEval(c);
+    if tryEvalSuccess then seq(r, nullE)
```
Keep in mind `tryEvalSucess` can be true with certain errors. e.g.
```m2
i6 : class \ trap (1 + 1; return)

o6 = (Error, Nothing)

o6 : Sequence

i7 : class \ trap (1 + 1; step)

o7 = (Nothing, Error)

o7 : Sequence

i8 : class \ trap (1 + 1; continue)

o8 = (Error, Nothing)

o8 : Sequence

i9 : class \ trap (1 + 1; break)

o9 = (Error, Nothing)

o9 : Sequence

i10 : trap (1 + 1; break)

--error or time limit reached in conversion of output to net: type 'debugError()' to run it again; will try conversion to string

--error in conversion of output to string
stdio:10:0:(3): error: at print: no method found for applying between to:
     argument 1 :  "," (of class String)
     argument 2 :  null (of class Nothing)
```
Perhaps we should rename this variable to avoid confusion?

mahrud replied:
Maybe it should be `tryCaughtError` which is true if a catchable error is caught and false if there's either no error or the error is not catchable (e.g. interrupts, syntax errors, etc).

d-torrance replied:
I like it!

mahrud replied:
I think the new commit still has this behavior, no?
```m2
i6 : class \ trap (1 + 1; return)

o6 = (Error, Nothing)
```

mahrud replied:
I think in this case we should directly return `err`, meaning something like:
```m2
i1 : class \ trap (1 + 1; return)
stdio:1:12:(3): error: unhandled return command
```
(i.e. I believe `trap` shouldn't catch syntax errors, just signaled errors)

d-torrance replied:
Oops, yeah.  ~~I blindly just swapped true and false without thinking!!~~ Actually, it never did this.  But it should!

d-torrance committed:
Rename tryEvalSuccess -> tryCaughtError for clarity

Also swap its values, i.e.:

* true = yes, we caught an error
* false = no, we didn't, either because it wasn't an error or it was
  some error we couldn't catch (syntax, interrupting)

d-torrance committed:
Don't trap errors that aren't caught by "trap"

Raise them instead

d-torrance committed:
Update "disassemble" tests for new try behavior

mahrud commented on M2/Macaulay2/d/actors4.d:
```diff
@@ -992,10 +985,13 @@ tostringfun(e:Expr):Expr := (
+     is x:List do (
+	 if ancestor(x.Class, filePositionClass)
+	 then toExpr(tostringFilePosition(e))
+	 else toExpr("<<a list>>"))
```
Not to add to your work but would it be simpler if FilePosition was its own interpreter-level class?

d-torrance replied:
That makes a lot of sense -- we could probably just add `Position` to the `Expr` union and use that.  Maybe in another PR, though.

mahrud committed:
moved trapfun close to tryEval

mahrud committed:
separated evalTryCode in evaluate.d

mahrud committed:
added 'try .. then .. except .. do ..' clause

mahrud commented:
I think this PR is good to go. While testing I found some pre-existing issues with robust error printing, but I'll make those as separate issues.

mahrud commented:
@d-torrance could you find an example of an error that is time limited by the new robust printing code? I just want to make sure everything is working as intended.

d-torrance commented:
It works!

```m2
i1 : Foo = new Type of HashTable;

i2 : net Foo := toExternalString Foo := x -> (sleep 10; "foo");

i3 : x = new HashTable; x#(new Foo)
stdio:3:20:(3): error: key not found in hash table:
        new Foo from {} (of class Foo)

i5 : error new Foo
stdio:4:5:(3): error: new Foo from {}
```

mahrud commented:
Why does the first one take 6 seconds?
```m2
i1 : Foo = new Type of HashTable;

i2 : net Foo := toExternalString Foo := x -> (sleep 10; "foo");

i3 : x = new HashTable;

i4 : elapsedTime x#(new Foo)
stdio:4:13:(3): error: key not found in hash table:
        new Foo from {} (of class Foo)
 -- 6.00091s elapsed

i5 : elapsedTime error new Foo
stdio:5:17:(3): error: new Foo from {}
 -- 3.00045s elapsed
```

d-torrance commented:
`silentRobustNet` tries both `net` and `toExternalString` before it resorts to `simpleToString`.  In this silly example, we're sleeping 3 seconds for each one

mahrud commented:
Maybe I've misunderstood the robust printing things, but shouldn't `toString` also be capped at 3s?
```m2
i9 : apply({net, toString, toExternalString}, m -> m Foo := x -> elapsedTime (print toString m; sleep 10; "foo"));

i10 : elapsedTime x#(new Foo)
net
 -- 3.00035s elapsed
toExternalString
 -- 3.0002s elapsed
toString
 -- 10.0005s elapsed
stdio:10:13:(3): error: key not found in hash table:
        foo (of class Foo)
 -- 16.0019s elapsed
```

d-torrance commented:
Where is it calling `toString`?  That's weird...

pzinn commented:
can we have a summary of what this PR currently does? it seems to have diverged quite a bit from the original one.

d-torrance commented:
> can we have a summary of what this PR currently does? it seems to have diverged quite a bit from the original one.

Great idea!  Here's what we've got:

* A new top-level (and subclass-able) `Error` type, implemented as a `SpecialExpr` containing an interpreter-level `Error`.  Access the message and position using `toString` and `locate`, resp.
* As the PR title says, a new `trap` keyword for Go-style error handling, i.e., one that returns a two-element sequence `(val, err)`, where `val` is the return value and `err` is null unless an error occured, in which case `val` is null and `err` is an `Error` object.
* New `try ... except ... do` and `try ... then ... except ... do` statements.  Inside the `do` clause, we can refer to the symbol specified by `except` to get the `Error` object corresponding to an error that occurrred, e.g. `try 1/0 except err do err` will return a "division by zero" `Error` object.
* Updates to `error`, including:
  * We can call `error` on an `Error` object to raise the corresponding error.
  * It is now completely defined in the interpreter, which should improve stack traces.
  * The helper functions used by `error` (`processArgs` and robust printing of objects) have also been moved to the interpreter

pzinn commented:
OK, my only worry now is that it conflicts with my tentative error upgrade (which is partially implemented on the branch running on the Macaulay2Web server). So, since we're doing fairly extensive changes, let me propose mine:
can we make errors be arbitrary expressions, rather than strings? at the end of the day, one can still use `toString` or `net` or whatever to output it, but it leaves the possibility to use other means of output for non Standard modes.

d-torrance commented:
Rather than storing the message as an `Expr` and post-processing it into a string based on the top level mode, wouldn't it be equivalent to keep it as a `string` but pre-process the input to `error`?  This is essentially what we're doing now with the `processArgs` helper function, but we could make this mode based.  And we wouldn't need to touch all the low-level error-printing code that doesn't know what an `Expr` is yet.

So maybe `error` could look up something like `(class x)#{topLevelMode, PrintError}` and use it to convert its input into a string to store as the message.

d-torrance commented:
I played around with the idea of a mode-based `ErrorPrint` method a bit on another branch: https://github.com/d-torrance/M2/commit/9ae759d0e888d769f73bbd37b1a5fc73f6f87a58

I think it's a bit cleaner, as we don't have to worry about another robust printing method in the interpreter.

*Edit:*  New version with an `AfterErrorPrint` method so we can handle the "here is the first use of..." messages for symbols more cleanly: https://github.com/d-torrance/M2/commit/b18f6e470b3b4d9c57edaf8f29528a9d1c9aff4d

mahrud commented:
> can we make errors be arbitrary expressions, rather than strings?

Unless I'm misunderstanding, I think this is already possible with this PR, eg:
```m2
i2 : new Error from PARA { "oops" }

o2 = PARA{"oops"}

o2 : Error
```
I guess the remaining part is adjusting `processErrorArgs0` in `debugging.dd` to not call `robustPrintString`, but something like `robustPrintError` which calls something like `Thing.RobustPrintErrorMethod` defined in `robust.m2` and can be adjusted based on mode.

I suspect this should be straightforward to do on top of this PR, but my preference is to finalize this PR first.

pzinn commented:
> Rather than storing the message as an `Expr` and post-processing it into a string based on the top level mode, wouldn't it be equivalent to keep it as a `string` but pre-process the input to `error`? This is essentially what we're doing now with the `processArgs` helper function, but we could make this mode based. And we wouldn't need to touch all the low-level error-printing code that doesn't know what an `Expr` is yet.
> 
> So maybe `error` could look up something like `(class x)#{topLevelMode, PrintError}` and use it to convert its input into a string to store as the message.

I think that would not be so useful -- most error messages come from the interpreter, not from explicit calls to `error`.

mahrud commented:
I presume this is just a rebase?

d-torrance commented:
Yup -- fixing the conflicct

mahrud commented:
@antonleykin @pzinn @mikestillman any objections to merging this for now, with the expectation that there may be future changes as Paul indicated? (And, of course, the possibility of reverting one or both new features before the release if we discover major problems)

pzinn commented:
fine with me

mikestillman commented:
I am looking at the final PR, and will review it too, and I will have comments (already do, but going back over the long discussion to make sure I don't re-ask too many questions.  I think I like this version now, but I want to make sure it doesn't slow things down, and that the doc matches the actual behavior (e.g. doc mentions `try code except code else code` but this isn't allowed?)

d-torrance commented:
Oops -- `try x except y else z` was definitely a typo!  I've fixed it, and also documented the `try/then/except/do` case that Mahrud added.

mikestillman commented:
@d-torrance Thanks!  Perhaps we should also add to the "error handling" node "stopIfError" (a mutable variable), and a sentence about `trap`?

It seems like other things that I thought were a problem have somehow been fixed since I tried this yesterday? Not sure... (I needed to do `new SelfInitializingType of Error` rather than `new Type of Error`.  Both seem to work now...)

d-torrance committed:
Add a couple things to "error handling" documentation node

Mention "trap" in the description, and also add a link to stopIfError.

d-torrance commented:
Sounds good!  I've updated "error handling" as requested.

The only difference between `SelfInializingType` and `Type`, I think, is that the former lets you do `MyError "foo"` but the latter needs `new MyError from "foo"`.

mikestillman commented:
It seems like we have 5 different forms for try blocks.  Perhaps we should deprecate a couple versions, or give guidance about which should be used?  Which ones are currently used? (That question is for myself, I'll take a look!)

I do like the trap function/keyword though!

mikestillman commented:
I found the problem: The extended markdown package (well, one of 3 different packages I just disabled) was grabbing shift-enter...  Disabling them, and restarting makes it work.

d-torrance commented:
There are six forms:
```m2
try x
try x else z
try x then y
try x then y else z
try x except err do z
try x then y except err do z
```

They're all documented in this PR.   I don't see a problem with having this many -- it doesn't hold a candle to all the different forms of `for` statements, for example!

mahrud commented:
I agree with Doug about different forms of try.

> The extended markdown package

There's a markdown package?!

mahrud commented:
@mikestillman is this good to merge?

d-torrance commented:
Discussed with Mike
